File: 1-50/problem7/Problem7.java
import java.util.ArrayList;
import java.util.List;

public class Problem7 {
    public static void main(String[] args) {
        final int MAX_COUNT = 10001;
        List<Long> primes = new ArrayList<Long>(10001);

        int count = 1;
        long next = 3;
        primes.add(2L);
        while (count < MAX_COUNT) {
            boolean isPrime = true;
            long max = (long)Math.sqrt(next);
            for (long i: primes) {
                if (max < i) break;
                if (next % i == 0) {
                    isPrime = false;
                    break;
                }
            }

            if (isPrime) {
                count++;
                primes.add(next);
            }

            next+=2;
        }

        System.out.println(next - 2);
    }
}

File: 1-50/problem7/Problem7.scala
val primes = scala.collection.mutable.ListBuffer(2)
var count = 1
var p = 3

while (count < 10001) {
    var isPrime = true
    var max = math.sqrt(p)
    var iter = primes.iterator
    var i = 0
    while (iter.hasNext && isPrime && max > i) {
        i = iter.next
        if (p % i == 0) isPrime = false
    }
    if (isPrime) {
        primes.append(p)
        count += 1
    }
    p += 2
}
println(primes.last)